delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala

```diff
@@ -0,0 +1,1088 @@
+   * 3. Calculating unmodified data by reading the touched files again and do anti join against
+   * output of [[JoinedRowProcessor]] using `__metadata_file_path` and `row_id`. Here we avoid
+   * shuffle of target files with too methods:
+   *  a. Pushing down a flag to [[org.apache.spark.sql.execution.FileSourceScanExec]] to enforce
+   *  one file per partition.
+   *  b. Adding [[RapidsRepartitionByFilePath]] to repartition the data by `__metadata_file_path`
+   *  without shuffle.
+   */
+  class LowShuffleMergeExecutor(
+      override val spark: SparkSession,
+      override val deltaTxn: OptimisticTransaction,
+      override val cmd: GpuLowShuffleMergeCommand) extends MergeExecutor {
+
+    // We over-count numTargetRowsDeleted when there are multiple matches;
+    // this is the amount of the overcount, so we can subtract it to get a correct final metric.
+    private var multipleMatchDeleteOnlyOvercount: Option[Long] = None
+    private val filesToRewrite = this.findTouchedFiles()
+
+
+    // UDFs to update metrics
+    // Make UDFs that appear in the custom join processor node deterministic, as they always
+    // return true and update a metric. Catalyst precludes non-deterministic UDFs that are not
+    // allowed outside a very specific set of Catalyst nodes (Project, Filter, Window, Aggregate).
+    private val incrSourceRowCountExpr = cmd.makeMetricUpdateUDF("numSourceRowsInSecondScan",
+      deterministic = true)
+
+    // Apply an outer join to find both, matches and non-matches. We are adding two boolean fields
+    // with value `true`, one to each side of the join. Whether this field is null or not after
+    // the outer join, will allow us to identify whether the resultant joined row was a
+    // matched inner result or an unmatched result with null on one side.
+    // We add row IDs to the targetDF if we have a delete-when-matched clause with duplicate
+    // matches and CDC is enabled, and additionally add row IDs to the source if we also have an
+    // insert clause. See above at isDeleteWithDuplicateMatchesAndCdc definition for more details.
+    private val sourceDF = cmd.getSourceDF(spark)
+      .withColumn(SOURCE_ROW_PRESENT_COL, new Column(incrSourceRowCountExpr))
+
+    private val targetDF = {
+      // Generate a new target dataframe that has same output attributes exprIds as the target plan.
+      // This allows us to apply the existing resolved update/insert expressions.
+      val baseTargetDF = buildTargetDFWithFiles(filesToRewrite)
+
+      // Here we enforce that the targetDF to read each file per partition. This is to ensure
+      // that the `_metadata_row_id` column is correctly populated for each row in the targetDF.
+      val newPlan = baseTargetDF.queryExecution.analyzed.transform {
+        case r@LogicalRelation(fs: HadoopFsRelation, _, _, _) =>
+          val newFs = fs.copy(options = fs.options +
+            (RapidsConf.FORCE_ONE_FILE_PER_PARITION.key -> "true"))(spark)
```

Comment:
What if the scan does not run on the GPU?  The user can explicitly disable the scan via configs, or there may be some datatype that we cannot load but otherwise can "ride along" in dataframes.  Seems like there needs to be additional tagging logic to catch this and fallback or otherwise handle it properly.  Also needs tests for this corner case.